packages/phpstan-rules/src/Rules/CheckClassNamespaceFollowPsr4Rule.php

```diff
@@ -0,0 +1,154 @@
+        $shortClassName = (string) $shortClassName;
+        $file = str_replace('\\', '/', $scope->getFile());
+        $namespaceBeforeClass = substr($namespacedName, 0, - strlen($shortClassName));
+
+        foreach ($this->psr4 as $namespace => $directory) {
+            $namespace = rtrim($namespace, '\\') . '\\';
+            if ($namespaceBeforeClass === $namespace) {
+                return [];
+            }
+
+            if (! $this->isInDirectoryNamed($scope, $directory)) {
+                continue;
+            }
+
+            if ($this->isClassNamespaceCorrect($namespace, $directory, $namespaceBeforeClass, $file)) {
+                return [];
+            }
+        }
+
+        $type = $this->getType($namespacedName, $file);
+        return [sprintf(self::ERROR_MESSAGE, $type, substr($namespaceBeforeClass, 0, -1))];
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(self::ERROR_MESSAGE, [
+            new CodeSample(
+                <<<'CODE_SAMPLE'
+// defined "Foo\\Bar" in composer.json
+namespace Foo;
+
+class Baz
+{
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+// defined "Foo\\Bar" in composer.json
+namespace Foo\Bar;
+
+class Baz
+{
+}
+CODE_SAMPLE
+            ),
+        ]);
+    }
+
+    private function getType(string $namespacedName, string $file): string
+    {
+        // totally different namespace needs include file
+        include_once $file;
+
+        $type = 'Class';
+
+        if (trait_exists($namespacedName)) {
+            $type = 'Trait';
+        }
+
+        if (interface_exists($namespacedName)) {
+            $type = 'Interface';
+        }
+
+        return $type;
+    }
+
+    private function isClassNamespaceCorrect(
+        string $namespace,
+        string $directory,
+        string $namespaceBeforeClass,
+        string $file
+    ): bool {
+        /** @var array<int, string> $paths */
+        $paths = explode($directory, $file);
+        if (count($paths) === 1) {
+            return false;
+        }
+
+        $namespaceSuffixByDirectoryClass = ltrim(str_replace('/', '\\', dirname($paths[1])), '\\');
+        $namespaceSuffixByNamespaceBeforeClass = rtrim(
+            (string) substr($namespaceBeforeClass, strlen($namespace)),
+            '\\'
+        );
+
+        return $namespaceSuffixByDirectoryClass === $namespaceSuffixByNamespaceBeforeClass;
+    }
+
+    private function getPsr4Autoload(SmartFileSystem $smartFileSystem): array
+    {
+        $composerJsonFile = './composer.json';
+        if (! file_exists($composerJsonFile)) {
+            return [];
+        }
+
+        $composerJsonContent = json_decode($smartFileSystem->readFile($composerJsonFile), true);
+        $autoloadPsr4 = $composerJsonContent['autoload']['psr-4'] ?? [];
+        $autoloadDevPsr4 = $composerJsonContent['autoload-dev']['psr-4'] ?? [];
+
+        return $autoloadPsr4 + $autoloadDevPsr4;
+    }
```

Comment:
Could you extract this to own service? It might be re-used in other parts

Reply:
I will try.

Reply:
moved to new service `ComposerAutoloadResolver`